File: src/core/riskManager.ts
import { logger } from '../utils/logger';

export interface RiskConfig {
  maxPositionPercent: number; // % от баланса (например, 10 = 10%)
  stopLossPercent: number; // % убытка (например, 2 = 2%)
  takeProfitPercent: number; // % прибыли (например, 5 = 5%)
  maxDailyTrades: number; // Максимум сделок в день
  minTradeInterval: number; // Минимальный интервал между сделками (секунды)
  trailingStopEnabled: boolean; // Включить trailing stop
  trailingStopPercent?: number; // % для trailing stop
}

export interface TradeSize {
  amountInCurrency: number; // Количество криптовалюты
  amountInUSDT: number; // Размер в USDT
  stopLoss: number; // Цена stop loss
  takeProfit: number; // Цена take profit
}

export class RiskManager {
  private config: RiskConfig;
  private dailyTradeCount: number = 0;
  private lastTradeTime: number = 0;
  private dailyTradeDate: string = '';

  constructor(config?: Partial<RiskConfig>) {
    this.config = {
      maxPositionPercent: config?.maxPositionPercent || 10,
      stopLossPercent: config?.stopLossPercent || 2,
      takeProfitPercent: config?.takeProfitPercent || 5,
      maxDailyTrades: config?.maxDailyTrades || 10,
      minTradeInterval: config?.minTradeInterval || 300, // 5 минут
      trailingStopEnabled: config?.trailingStopEnabled || false,
      trailingStopPercent: config?.trailingStopPercent || 1
    };

    logger.info('Risk Manager initialized:', this.config);
  }

  /**
   * Рассчитать размер позиции
   */
  calculatePositionSize(
    balance: number,
    currentPrice: number,
    side: 'buy' | 'sell'
  ): TradeSize {
    // Размер позиции в USDT
    const positionSizeUSDT = balance * (this.config.maxPositionPercent / 100);

    // Количество криптовалюты
    const amountInCurrency = positionSizeUSDT / currentPrice;

    // Рассчитываем Stop Loss и Take Profit
    let stopLoss: number;
    let takeProfit: number;

    if (side === 'buy') {
      // Для long позиции
      stopLoss = currentPrice * (1 - this.config.stopLossPercent / 100);
      takeProfit = currentPrice * (1 + this.config.takeProfitPercent / 100);
    } else {
      // Для short позиции
      stopLoss = currentPrice * (1 + this.config.stopLossPercent / 100);
      takeProfit = currentPrice * (1 - this.config.takeProfitPercent / 100);
    }

    logger.debug(`Position size calculated: ${amountInCurrency.toFixed(6)} @ ${currentPrice}`);
    logger.debug(`Stop Loss: ${stopLoss.toFixed(2)}, Take Profit: ${takeProfit.toFixed(2)}`);

    return {
      amountInCurrency,
      amountInUSDT: positionSizeUSDT,
      stopLoss,
      takeProfit
    };
  }

  /**
   * Проверить, можно ли открыть новую сделку
   */
  canOpenTrade(): { allowed: boolean; reason?: string } {
    // Проверка дневного лимита
    const today = new Date().toISOString().split('T')[0];
    if (this.dailyTradeDate !== today) {
      this.dailyTradeCount = 0;
      this.dailyTradeDate = today;
    }

    if (this.dailyTradeCount >= this.config.maxDailyTrades) {
      return {
        allowed: false,
        reason: `Daily trade limit reached (${this.config.maxDailyTrades})`
      };
    }

    // Проверка минимального интервала
    const now = Date.now();
    const timeSinceLastTrade = (now - this.lastTradeTime) / 1000;

    if (this.lastTradeTime > 0 && timeSinceLastTrade < this.config.minTradeInterval) {
      return {
        allowed: false,
        reason: `Min interval not met (${timeSinceLastTrade.toFixed(0)}s < ${this.config.minTradeInterval}s)`
      };
    }

    return { allowed: true };
  }

  /**
   * Зарегистрировать открытие сделки
   */
  registerTrade(): void {
    this.dailyTradeCount++;
    this.lastTradeTime = Date.now();
    logger.info(`Trade registered. Daily count: ${this.dailyTradeCount}/${this.config.maxDailyTrades}`);
  }

  /**
   * Проверить, нужно ли закрыть позицию (Stop Loss / Take Profit)
   */
  shouldClosePosition(
    _entryPrice: number, // Префикс _ означает, что параметр намеренно не используется
    currentPrice: number,
    side: 'long' | 'short',
    stopLoss: number,
    takeProfit: number
  ): { shouldClose: boolean; reason?: string } {
    if (side === 'long') {
      // Long позиция
      if (currentPrice <= stopLoss) {
        return { shouldClose: true, reason: 'Stop Loss hit' };
      }
      if (currentPrice >= takeProfit) {
        return { shouldClose: true, reason: 'Take Profit hit' };
      }
    } else {
      // Short позиция
      if (currentPrice >= stopLoss) {
        return { shouldClose: true, reason: 'Stop Loss hit' };
      }
      if (currentPrice <= takeProfit) {
        return { shouldClose: true, reason: 'Take Profit hit' };
      }
    }

    return { shouldClose: false };
  }

  /**
   * Рассчитать прибыль/убыток
   */
  calculateProfitLoss(
    entryPrice: number,
    currentPrice: number,
    amount: number,
    side: 'long' | 'short'
  ): { profit: number; profitPercent: number } {
    let profit: number;

    if (side === 'long') {
      profit = (currentPrice - entryPrice) * amount;
    } else {
      profit = (entryPrice - currentPrice) * amount;
    }

    const profitPercent = (profit / (entryPrice * amount)) * 100;

    return { profit, profitPercent };
  }

  /**
   * Обновить конфигурацию
   */
  updateConfig(config: Partial<RiskConfig>): void {
    this.config = { ...this.config, ...config };
    logger.info('Risk Manager config updated:', this.config);
  }

  /**
   * Получить текущую конфигурацию
   */
  getConfig(): RiskConfig {
    return { ...this.config };
  }

  /**
   * Получить статистику
   */
  getStats(): {
    dailyTradeCount: number;
    maxDailyTrades: number;
    lastTradeTime: number;
  } {
    return {
      dailyTradeCount: this.dailyTradeCount,
      maxDailyTrades: this.config.maxDailyTrades,
      lastTradeTime: this.lastTradeTime
    };
  }
}

File: backend_py/core/risk_manager.py
from typing import Dict, Any
from datetime import datetime
from loguru import logger
from pydantic import BaseModel

class RiskConfig(BaseModel):
    max_position_percent: float = 10  # % от баланса
    stop_loss_percent: float = 2  # % убытка
    take_profit_percent: float = 5  # % прибыли
    max_daily_trades: int = 10  # Максимум сделок в день
    min_trade_interval: int = 300  # Минимальный интервал между сделками (секунды)
    trailing_stop_enabled: bool = False
    trailing_stop_percent: float = 1
    max_daily_loss_percent: float = 5
    max_drawdown_percent: float = 10
    min_order_size: float = 10
    max_order_size: float = 1000

class TradeSize(BaseModel):
    amount_in_currency: float  # Количество криптовалюты
    amount_in_usdt: float  # Размер в USDT
    stop_loss: float  # Цена stop loss
    take_profit: float  # Цена take profit

class RiskManager:
    def __init__(self, config: Dict[str, Any]):
        self.config = RiskConfig(**config)
        self.daily_trade_count = 0
        self.last_trade_time = 0
        self.daily_trade_date = ''
        self.daily_loss = 0

        logger.info(f'Risk Manager initialized: {self.config}')

    def calculate_position_size(self, balance: float, current_price: float) -> float:
        """Рассчитать размер позиции в USDT"""
        position_size_usdt = balance * (self.config.max_position_percent / 100)

        # Ограничиваем минимальным и максимальным размером
        position_size_usdt = max(self.config.min_order_size, min(position_size_usdt, self.config.max_order_size))

        return position_size_usdt

    def calculate_stop_loss(self, entry_price: float, side: str) -> float:
        """Рассчитать цену стоп-лосс"""
        if side == 'long':
            return entry_price * (1 - self.config.stop_loss_percent / 100)
        else:  # short
            return entry_price * (1 + self.config.stop_loss_percent / 100)

    def calculate_take_profit(self, entry_price: float, side: str) -> float:
        """Рассчитать цену тейк-профит"""
        if side == 'long':
            return entry_price * (1 + self.config.take_profit_percent / 100)
        else:  # short
            return entry_price * (1 - self.config.take_profit_percent / 100)

    def can_open_position(self, current_time: float) -> bool:
        """Проверить, можно ли открыть позицию"""
        # Проверяем дневной лимит сделок
        today = datetime.now().strftime('%Y-%m-%d')
        if self.daily_trade_date != today:
            self.daily_trade_count = 0
            self.daily_trade_date = today

        if self.daily_trade_count >= self.config.max_daily_trades:
            logger.warning('Daily trade limit reached')
            return False

        # Проверяем интервал между сделками
        if current_time - self.last_trade_time < self.config.min_trade_interval:
            logger.warning('Trade interval too short')
            return False

        # Проверяем дневной убыток
        if self.daily_loss >= self.config.max_daily_loss_percent:
            logger.warning('Daily loss limit reached')
            return False

        return True

    def record_trade(self, profit_loss: float = 0):
        """Записать сделку"""
        self.daily_trade_count += 1
        self.last_trade_time = datetime.now().timestamp()
        self.daily_loss += max(0, -profit_loss)  # Только убытки

        logger.info(f'Trade recorded: count={self.daily_trade_count}, daily_loss={self.daily_loss:.2f}%')